File: Code-Chef/contests/COOK01/RIPPLE.cpp
#include<bits/stdc++.h>
using namespace std;

string sub(string &a,string &b)
{
    int l=a.size(),borrow=0,c,d;
    string res(l,'0');
    for(int i=l-1;i>=0;i--){
        c=(a[i]-'0')-borrow;
        d=b[i]-'0';
        borrow=0;
        if(c-d>=0)
            res[i]=c-d+'0';
        else{
            borrow=1;
            res[i]=2+(c-d)+'0';
        }
    }
    return res;
}

string add(string &a,string &b)
{
    int l=a.size(),carry=0,c,d;
    string res(l,'0');
    for(int i=l-1;i>=0;i--){
        c=(a[i]-'0')+carry;
        d=b[i]-'0';
        carry=0;
        if(c+d>1)
        {
            res[i]=(c+d)%2+'0';
            carry=(c+d)/2;
        }
        else{
            res[i]=c+d+'0';
        }
    }
    if(carry)
    {
        res='1'+res;        
    }
    return res;
}

string bin(int val)
{
    string s;
    while(val)
    {
        s.push_back(val%2+'0');
        val/=2;
    } 
    reverse(s.begin(),s.end());
    // cout<<"reverse:"<<s<<endl;
    return s;
}

int main()
{
    int t,n,x=0,y=0;
    string a,b,result,z;
    cin>>t;
    while(t--){
        cin>>n;
        cin>>a>>b;
        a+='0';b+='0';
        x=0,y=0;
        for(int i=0;i<n;i++) if(a[i]=='1') x++;
        for(int i=0;i<n;i++) if(b[i]=='1') y++;
        result=sub(b,a);
        if(y-x>0){
            z=bin(y-x);
            z=string(a.size()-z.size(),'0')+z;
            result=sub(result,z);
        }
        else
        {
            z=bin(x-y);
            z=string(a.size()-z.size(),'0')+z;
            result=add(result,z);
        }
        auto pp=result.find('1'); 
        if(pp!=string::npos)  cout<<result.substr(pp);

    }
    return 0;
}

File: Code-Chef/contests/COOK01/RIPPLE.py
t=int(input())
def sub(a,b):
    l=len(a)
    res=''
    borrow=0
    for i in range(l-1,-1,-1):
        c,d=int(a[i]),int(b[i])
        c=c-borrow
        borrow=0
        if c-d>=0:
            res=str(c-d)+res
        else:
            borrow=1
            res=str(2+(c-d))+res
    return res

def add(a,b):
    l=len(a)
    res=''
    carry=0
    for i in range(l-1,-1,-1):
        c,d=int(a[i]),int(b[i])
        c=c+carry
        carry=0
        if c+d>1:
            res=str((c+d)%2)+res
            carry=(c+d)//2
        else:
            res=str(c+d)+res
    return res

while(t):
    n=int(input())
    a=input()
    b=input()
    a=a+'0'
    b=b+'0'
    x=a.count('1')
    y=b.count('1')
    result=sub(b,a)
    if(y-x>0):
        z=bin(y-x)[2:]
        z='0'*(len(a)-len(z))+z
        result=sub(result,z)
    else:
        z=bin(x-y)[2:]
        z='0'*(len(a)-len(z))+z
        result=add(result,z)
    print(result[result.find('1'):])
    t-=1